feat: assumeCorrectReference option resolves #2715

Heinrich-vanNieuwenhuizen opened the pull request:
**Summary**
assumeCorrectReference option added
assume the correct reference of the dart code in views, when the resolver fails.

**Description**
This will allow the ability to use drift tables and dart views. The resolver cannot resolve references of the tables as it doesn't have any context of it yet. 

But by coding ahead we know what those references are thus we can use type-safe views. 

**Example**
build.yaml
```yaml
targets:
  $default:
    builders:
      drift_dev:
        options:
          assume_correct_reference : true
```
descriptors.drift
```sql
CREATE TABLE [DESCRIPTORS](
	[DescriptorID] [int] NOT NULL PRIMARY KEY,
	[HierStrucID] [bigint] NULL,
	[DescType] [int] NULL,
	[LineNumber] [int] NULL,
	[StringNumberID] [bigint] NULL,
	[IsVisible] [bit] NULL,
	[IsDeleted] [bit] NULL,
	[OptionBits] [nvarchar](16) NULL);
```
string_table.drift
```sql
CREATE TABLE [STRING_TABLE](
	[StringID] [bigint] NOT NULL PRIMARY KEY,
	[StringNumberID] [bigint] NULL,
	[LangID] [int] NULL,
	[IsVisible] [bit] NOT NULL DEFAULT ((1)),
	[IsDeleted] [bit] NOT NULL DEFAULT ((0)),
	[HierStrucID] [bigint] NULL,
	[PosRef] [bigint] NULL,
	[StringText] [nvarchar](128) NULL
);
```
descriptor_view.dart
```dart
abstract class DescriptorView extends View {
  late final DatastoreDb attachedDatabase;
  IntColumn get descriptorID => attachedDatabase.descriptors.descriptorID;
  IntColumn get hierStrucID => attachedDatabase.descriptors.hierStrucID;
  IntColumn get descType => attachedDatabase.descriptors.descType;
  IntColumn get lineNumber => attachedDatabase.descriptors.lineNumber;
  RealColumn get isVisible => attachedDatabase.descriptors.isVisible;
  RealColumn get isDeleted => attachedDatabase.descriptors.isDeleted;
  TextColumn get optionBits => attachedDatabase.descriptors.optionBits;
  TextColumn get stringText => attachedDatabase.stringTable.stringText;

  @override
  Query as() => select([
    descriptorID,
    hierStrucID,
    descType,
    lineNumber,
    isVisible,
    isDeleted,
    optionBits,
    stringText
  ]).from(attachedDatabase.descriptors).join([
    innerJoin(
        attachedDatabase.stringTable,
        attachedDatabase.stringTable.stringNumberID
            .equalsExp(attachedDatabase.descriptors.stringNumberID)),
  ]);
}
```

Heinrich-vanNieuwenhuizen committed:
assumeCorrectReference resovels #2715

assumeCorrectReference option added
assume the correct reference of the dart code in views, when resolver fails.

simolus3 commented on drift_dev/test/backends/build/build_integration_test.dart:
```diff
@@ -614,6 +616,153 @@ class MyDatabase {
+        logger: debugLogger
+    );
+
+    // var actual = utf8.decode(result.writer.assets[(result.writer.assets.keys
+    //     .firstWhere((key) => key.path == ('lib/drift/datastore_db.drift.dart')))]!);
```
Leftover debugging code

Heinrich-vanNieuwenhuizen replied:
done

simolus3 commented on drift_dev/test/backends/build/build_integration_test.dart:
```diff
@@ -614,6 +616,153 @@ class MyDatabase {
+);
+	''',
+        },
+        modularBuild: true,
```
Wouldn't this actually generate invalid code in the modular build, since that tries to infer imports which isn't possible with this? Let's disable this for this test, that also avoids having to test for the two other outputs.

Heinrich-vanNieuwenhuizen replied:
Done, it did work but we don't require modularBuild

simolus3 commented on drift_dev/lib/src/analysis/results/view.dart:
```diff
@@ -96,6 +96,6 @@ class DartViewSource extends DriftViewSource {
   final List<TableReferenceInDartView> staticReferences;
-
-  DartViewSource(this.dartQuerySource, this.primaryFrom, this.staticReferences);
+  final String? staticSource;
```
There should be a comment explaining that this is used when referring to an element not yet generated.

Heinrich-vanNieuwenhuizen replied:
Done

simolus3 commented on drift_dev/lib/src/analysis/resolver/dart/view.dart:
```diff
@@ -209,7 +227,7 @@ class DartViewResolver extends LocalElementResolver<DiscoveredDartView> {
         final reference =
             references.firstWhereOrNull((ref) => ref.name == parts[0]);
-        if (reference == null) {
+        if (reference == null || reference.table == null) {
```
Is this a leftover from a previous iteration? `TableReferenceInDartView.table` is a non-nullable field.

Heinrich-vanNieuwenhuizen replied:
Done. Yes not sure what I tried there.

simolus3 commented on drift_dev/lib/src/analysis/resolver/dart/view.dart:
```diff
@@ -188,11 +189,28 @@ class DartViewResolver extends LocalElementResolver<DiscoveredDartView> {
 
-    final query = AnnotatedDartCode.build(
-        (builder) => builder.addAstNode(body.expression, exclude: {target!}));
+    // if(resolver.driver.options.assumeCorrectReference){
+    //   bool separate = false;
+    //   for (int i = 0; i < query.elements.length; i++) {
+    //     if (separate && query.elements[i] is String && i != query.elements.length-1) {
+    //       query.elements[i]+=',';
+    //       separate = false;
+    //     }
+    //     separate = query.elements[i] is DartTopLevelSymbol;
+    //   }
+    // }
```
I don't understand what this did, is this solving another problem introduced by the option? Either way, if it's not used it should be removed.

Heinrich-vanNieuwenhuizen replied:
Done, was the first attempt as the joins did not add the separator "," . But then having a group by I could not use this simple for loop and resolved it with the current solution.

simolus3 reviewed (commented):
I'm not a big fan of this option since it essentially bypasses the analyzer, but I think we can have it for this specific use-cases since this used to work.

Heinrich-vanNieuwenhuizen committed:
Ready for merge

Removed residual code and comments.
Commented and explained staticSource
Removed modularBuild and tested test

simolus3 reviewed (approved):
Thanks!

Heinrich-vanNieuwenhuizen committed:
Apply dart formatting

Heinrich-vanNieuwenhuizen committed:
unused_import

Heinrich-vanNieuwenhuizen committed:
Apply dart formatting

Heinrich-vanNieuwenhuizen committed:
Merge branch 'simolus3:develop' into assume_correct_reference
